rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,188 @@
+```
+NT_Bool -> NT_Bool && NT_Bool | NT_int == NT_int 
+            | NT_int <= NT_int | NT_int < NT_int 
+            | SAME_OBJECT(terminals_ptr, terminals_ptr)
+            
+NT_int  -> NT_int + NT_int | terminals_int | LOOP_ENTRY(terminals_int)
+            | POINTER_OFFSET(terminals_ptr) | OBJECT_SIZE(terminals_ptr)
+            | POINTER_OFFSET(LOOP_ENTRY(terminals_ptr)) | 1
+```
+where `terminals_ptr` are all pointer variables in the scope, and `terminal_int` are all integers variables in the scope.
+
+We use the following example to illustrate how the synthesizer works.
+```rust
+#[kani::proof]
+fn main() {
+    let mut y: i32 = kani::any_where(|i| *i > 0);
+
+    while y > 0 {
+        y = y - 1;
+    }
+    assert!(y == 0);
+}
+```
+As there is only one variable `y` in the scope, the grammar template above will be instantiated to the following grammar
+```
+NT_Bool -> NT_Bool && NT_Bool | NT_int == NT_int 
+            | NT_int <= NT_int | NT_int < NT_int 
+NT_int  -> NT_int + NT_int | y | LOOP_ENTRY(y) | 1
+```
+The synthesizer will enumerate candidates derived from `NT_Bool` in the following order.
+```
+y == y
+y == LOOP_ENTRY(y)
+y == 1
+...
+1 <= y + 1
+...
+```
+The synthesizer then verifies with CBMC if the candidate is an inductive invariant that can be used to prove the post-condition `y == 0`.
+For example, the candidate `y == y` is verified to be an inductive invariant, but cannot be used to prove the post-condition `y == 0`.
+The candidate `y == 1` is not inductive.
+The synthesizer rejects all candidates until it finds the candidate `1 <= y + 1`, which can be simplified to `y >= 0`.
+`y >= 0` is an inductive invariant that can be used to prove the post-condition.
+So the synthesizer will return `y >= 0` and apply it in the goto model to get `main_synthesized.goto`.
+Then Kani will call `cbmc` on `main_synthesized.goto` to verify the program with the synthesized loop contracts. 
+
+## Rationale and alternatives
+
+- Different candidate space.
+The candidate grammar introduced above now only contains a restricted set of operators, which works well for array-manipulating programs with only pointer-checks instrumented by `goto-instrument`, but probably not enough for other user-written checks.
+We may want to include array-indexing, pointer-dereference, or other arithmetic operators in the candidate grammar for synthesizing a larger set of loop invariants.
+However, there is a trade-off between the size of candidate we enumerate and the running time of the enumeration.
+We will collect more data to decide what operators we should include in the candidate grammar.
+
+## Open questions
+
+**How does the synthesizer work with unwinding numbers?**
+There may exist some loops for which the synthesizer cannot find loop contracts, but some small unwinding numbers are enough to cover all executions of the loops.
+In this case, we may want to unwind some loops in the program while synthesizing loop contracts for other loops.
```

Comment:
It would be nice if we can introduce a macro to enable/disable unwinding, e.g.
```rust
#[kani::proof]
#[kani::unwind(10)]
fn check() {
    // unwinding starts as enabled, so all loops in this code block will be unwound to 10
    #[kani::disable_unwinding]
    // unwinding is disabled for all loops in this block of code
    #[kani::enable_unwinding]
    // it is enabled in this block of code until the end of the program
}
```
but not sure if this is feasible since it may require incorporating the synthesizer within the symbolic execution engine.

Reply:
Good point, in C we similarly use pragmas to enable/disable certain checks for a block of code.
What I am not sure about this is how common it is where we want to unwind some loops while synthesize loop contracts for others.
I will add it as future possibilities.